[lldb] Add utility to create Mach-O corefile from YAML desc

jasonmolenda opened the pull request:
I've wanted a utility to create a corefile for test purposes given a bit of memory and regsters, for a while.  I've written a few API tests over the years that needed exactly this capability -- we have several one-off Mach-O corefile creator utility in the API testsuite to do this.  But it's a lot of boilerplate when you only want to specify some register contents and memory contents, to create an API test.

This adds yaml2mach-core, a tool that should build on any system, takes a yaml description of register values for one or more threads, optionally memory values for one or more memory regions, and can take a list of UUIDs that will be added as LC_NOTE "load binary" metadata to the corefile so binaries can be loaded into virtual address space in a test scenario.

The format of the yaml file looks like

```
cpu: armv7m

# optionally specify the number of bits used for addressing
# (this line is from a different, 64-bit, yaml file)
addressable-bits: 
    num-bits: 39

# optionally specify one or more binary UUID and slide/virtual address to be added as an LC_NOTE
# (this line is from a different, 64-bit, yaml file)
binaries:
  - name: debug-binary.development
    uuid: 67942352-5857-3D3D-90CB-A3F80BA67B04
    virtual-address: 0xfffffff01840c000

threads:
  - regsets: 
     - flavor: gpr 
        registers: [{name: sp, value: 0x2000fe70}, {name: r7, value: 0x2000fe80}, 
                    {name: pc, value: 0x0020392c}, {name: lr, value: 0x0020392d}]

memory-regions:
  # stack memory
  - addr: 0x2000fe70 
    UInt32: [ 0x0000002a, 0x20010e58, 0x00203923, 
              0x00000001, 0x2000fe88, 0x00203911, 
              0x2000ffdc, 0xfffffff9 ]
  # instructions of a function
  - addr: 0x203910 
     UInt8: [ 0xf8, 0xb5, 0x04, 0xaf, 0x06, 0x4c, 0x07, 0x49, 
              0x74, 0xf0, 0x2e, 0xf8, 0x01, 0xac, 0x74, 0xf0 ]
```

and that's all that is needed to specify a corefile where four register values are specified (the others will be set to 0), and two memory regions will be emitted.

The memory can be specified as an array of UInt8, UInt32, or UInt64, I anticipate that some of these corefiles may have stack values constructed manually and it may be simpler for a human to write them in a particular grouping of values.

I needed this utility for an upcoming patch for ARM Cortex-M processors, to create a test for the change.  I took the opportunity to remove two of the "trivial mach-o corefile" creator utilities I've written in the past, which also restricted the tests to only run on Darwin systems because I was using the system headers for Mach-O constant values.

rdar://110663219

jasonmolenda committed:
[lldb] Add utility to create Mach-O corefile from YAML desc

I've wanted a utility to create a corefile for test purposes given
a bit of memory and regsters, for a while.  I've written a few API
tests over the years that needed exactly this capability -- we have
several one-off Mach-O corefile creator utility in the API testsuite
to do this.  But it's a lot of boilerplate when you only want to
specify some register contents and memory contents, to create an
API test.

This adds yaml2mach-core, a tool that should build on any system,
takes a yaml description of register values for one or more threads,
optionally memory values for one or more memory regions, and can
take a list of UUIDs that will be added as LC_NOTE "load binary"
metadata to the corefile so binaries can be loaded into virtual
address space in a test scenario.

The format of the yaml file looks like

cpu: armv7m
endian: little
threads:
  - regsets:
      - flavor: gpr
        registers: [{name: sp, value: 0x2000fe70}, {name: r7, value: 0x2000fe80},
                    {name: pc, value: 0x0020392c}, {name: lr, value: 0x0020392d}]

memory-regions:
  - addr: 0x2000fe70
    UInt32: [
      0x0000002a, 0x20010e58, 0x00203923, 0x00000001,
      0x2000fe88, 0x00203911, 0x2000ffdc, 0xfffffff9
    ]
  - addr: 0x203910
    UInt8: [
      0xf8, 0xb5, 0x04, 0xaf, 0x06, 0x4c, 0x07, 0x49,
      0x74, 0xf0, 0x2e, 0xf8, 0x01, 0xac, 0x74, 0xf0
    ]

and that's all that is needed to specify a corefile where four register
values are specified (the others will be set to 0), and two memory
regions will be emitted.

The memory can be specified as an array of UInt8, UInt32, or UInt64,
I anticipate that some of these corefiles may have stack values
constructed manually and it may be simpler for a human to write
them in a particular grouping of values.

Accepting "endian" is probably a boondoggle that won't ever come
to any use, and honestly I don't 100% know what the correct byte
layout would be for a big endian Mach-O file any more.  In a RISC-V
discussion a month ago, it was noted that register byte layout will
be little endian even when there is a big endian defined format for
RV, so memory would be byteswapped but registers would not.  It may
have been better not to pretend to support this, but on the other
hand it might be neat to be able to generate a big endian test case
simply.

I needed this utility for an upcoming patch for ARM Cortex-M processors,
to create a test for the change.  I took the opportunity to remove two
of the "trivial mach-o corefile" creator utilities I've written in the
past, which also restricted the tests to only run on Darwin systems
because I was using the system headers for Mach-O constant values.

rdar://110663219

jasonmolenda committed:
Whitespace fix.

jasonmolenda committed:
more ws fix

jasonmolenda commented:
One thing I wasn't thrilled about with llvm's yaml MappingTraits parser was that I need to define register values like
```
       registers: [
           {name: x0, value: 0x0000000000000001}, {name: x1, value: 0x000000016fdff3c0},
            {name: x2, value: 0x000000016fdff3d0}, {name: x3, value: 0x000000016fdff510},
```

instead of a more natural style of `registers = { "x0": 0x1, "x1": 0x16fdff3c0, "x2": 0x16fdff3d0}` or so.  
At least I couldn't figure out how to do this.  It makes the yaml descriptions noisier than they really need to be.

jasonmolenda commented:
Ah, I included an implementation to ObjectFileJSON::SetLoadAddress in this PR, which isn't actually relevant to the new tool or use of it.  It's for the testcase I am writing for the Cortex-M backtrace, where I will have an ObjectFileJSON binary specification and I needed to load its sections in virtual address space for the backtrace to work.  I can pull this out of the PR and add it to that PR when I put it up in a bit if anyone would prefer this done cleanly.

jasonmolenda committed:
Revert the ObjectFileJSON changes, will land separately.

jasonmolenda commented:
I think @labath would point out that I'm doing an end-run around making a sufficient Mock Process capability, with memory and threads and symbols, to write unit tests.  @medismailben would point out that we could write a Scripted Process python script that would ingest this same information and vend a Process, just as well as using the corefile container for the information.  For that matter, a little gdb remote serial protocol stub written in python could present this same information as if it were a live process with threads, registers, and memory.  

Because I already had several mach-o corefile creator tools (and needed a new one each time I needed to test another part of the mach-o corefile reader part of lldb), it seemed most natural to go that route, to me.  

The most important part for me is the simplicity of taking a real world debug problem situation, live or corefile, which may involve giant binaries/corefile and cannot be used in a test for size or confidentiality reasons, but we can extract the core bits of registers and memory that are sufficient to show the issue being fixed.  We can't test issues dealing with debug info with this mechanism -- say something specific to firmware debugging that can't be replicated in a userland process -- but for a lot of memory-and-stack-and-register type bugs, I think this could be a handy tool.

jasonmolenda committed:
Correct my cmake details so the `check-lldb-api` target builds the
new yaml2macho-core tool.  Thanks to Felipe for the guidance.

jasonmolenda committed:
Seems like I need <cstdint> for
uint32_t et al, for some reason I thought
those were built-in with C++.

bulbazord reviewed (commented):
This is a really cool utility! I'm no expert, but I made a few suggestions inline. Thanks for sharing. :)

jasonmolenda committed:
Remove the endian specifier.
I'm not convinced that it was emitting big-endian Mach-O files
correctly, and until this is actually needed, there's no point
in carrying around dubious code.

jasonmolenda commented on lldb/tools/yaml2macho-core/Utility.cpp:
```diff
@@ -0,0 +1,57 @@
+}
+
+void add_uint64(const CoreSpec &spec, std::vector<uint8_t> &buf, uint64_t val) {
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
```
I played with the byte swapping code a bit, and in the end I'm not feeling confident that it is emitting BE corefiles correctly, it's more likely to confuse someone into thinking there is a bug in their code.  I'm pulling that feature out until there's an actual need for it.

jasonmolenda committed:
Use lldb's UUID class.

JDevlieghere commented on lldb/tools/yaml2macho-core/CoreSpec.h:
```diff
@@ -0,0 +1,55 @@
+//===-- CoreSpec.h --------------------------------------------------------===//
```
Nit: new files shouldn't use this ASCII art anymore: https://discourse.llvm.org/t/is-c-in-header-files-still-relevant/83124
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/tools/yaml2macho-core/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+add_lldb_tool(yaml2macho-core
+    main.cpp
+    yaml2corespec.cpp
```
Nit: filename suggestions. 
```suggestion
    yaml2macho.cpp
    CoreSpec.cpp
```

JDevlieghere commented on lldb/tools/yaml2macho-core/CoreSpec.h:
```diff
@@ -0,0 +1,55 @@
+  MemoryType type;
+  uint32_t size;
+  // One of the following formats.
+  std::vector<uint8_t> bytes;
+  std::vector<uint32_t> words;
+  std::vector<uint64_t> doublewords;
```
If they're mutually exclusive, you could do:
```suggestion
  using Bytes = std::vector<uint8_t>;
  using Words = std::vector<uint32_t>;
  using Doublewords = std::vector<uint64_t>;
  std::variant<Bytes, Words, Doublewords> data;
```
but this might complicate the YAML traits. We do this for protocol messages in DAP and MCP.

jasonmolenda replied:
yeah, I wasn't super thrilled with having these three, but wanted to maintain the input formatting for endian switching (which I then later abandoned lol).  At one point it was an anonymous union.  I ended up bailing on all of that and just having three members, only one of which is active; they're not accessed directly in many places.

JDevlieghere commented on lldb/tools/yaml2macho-core/LCNoteWriter.cpp:
```diff
@@ -0,0 +1,55 @@
+//===----------------------------------------------------------------------===//
+
+#include "lldb/Utility/UUID.h"
+
```
Nit: no newline between header includes, the order is handled by clang-format.
```suggestion
```

JDevlieghere commented on lldb/tools/yaml2macho-core/main.cpp:
```diff
@@ -0,0 +1,223 @@
+int main(int argc, char **argv) {
+
+  const char *const short_opts = "i:o:u:h";
+  const option long_opts[] = {{"input", required_argument, nullptr, 'i'},
```
This should use `cl::opt` (https://llvm.org/docs/CommandLine.html). Even if the current implementation is slightly simpler, it makes it harder to extend in the future. It's used by every llvm tool (that doesn't use tablegen).

JDevlieghere commented on lldb/tools/yaml2macho-core/LCNoteWriter.cpp:
```diff
@@ -0,0 +1,55 @@
+
+#include "llvm/BinaryFormat/MachO.h"
+
+void create_lc_note_binary_load_cmd(const CoreSpec &spec,
```
Surprised this lives in its own file? Why not merge this with utility?

jasonmolenda replied:
It's a tiny one method file today, but I can see me adding additional LC_NOTEs in the future.  So I wanted the "thread writer" file which emits LC_THREAD load commands, the "memory writer" file which emits LC_SEGMENTs and the blocks of memory, and the "LC_NOTE writer" which emits LC_NOTE load commands and bytes of the note command payloads.  It makes sense in my head for these three to be in their own files.

JDevlieghere commented on lldb/tools/yaml2macho-core/Utility.cpp:
```diff
@@ -0,0 +1,22 @@
+#include "Utility.h"
+#include "CoreSpec.h"
+
+void add_uint64(const CoreSpec &spec, std::vector<uint8_t> &buf, uint64_t val) {
```
This takes a `CoreSpec` but doesn't actually use it? Same for `add_uint32` below.

jasonmolenda replied:
Originally you could specify the Endianness in the YAML and I needed to pass that to the add_uint methods to fix byte ordering in the output file.  I eventually lost confidence that I was actually creating correctly-formatted Big Endian mach-o files and I removed that feature, forgot to clean up and remove this argument.

JDevlieghere reviewed (commented):
I know this is a simple tool, but it seems like adding a bit of organization could go a long way. Essentially, the tool consists of 3 parts:

1. A reader that takes YAML and creates an in-memory/intermediate representation (CoreSpec). 
2. A writer that takes a CoreSpec and emit a binary. 
3. The glue that holds (1) and (2) together as well as command-line parsing and I/O. 

If it were up to me, that's how I would structure this tool. I think that will make it a lot easier to understand and extend in the future.

jasonmolenda commented:
> I know this is a simple tool, but it seems like adding a bit of organization could go a long way. Essentially, the tool consists of 3 parts:
> 
> 1. A reader that takes YAML and creates an in-memory/intermediate representation (CoreSpec).
> 2. A writer that takes a CoreSpec and emit a binary.
> 3. The glue that holds (1) and (2) together as well as command-line parsing and I/O.
> 
> If it were up to me, that's how I would structure this tool. I think that will make it a lot easier to understand and extend in the future.

Thanks for all the comments, addressing them now.  

Yeah before I wrote this I didn't have a clear idea of what it would look like when finished (for some reason, it seems obvious now, but in the beginning there were some poor choices made and fixed along the way).  You could imagine someone making an ELF corefile output capability, for instance.  The YAML files I'm using as input is just a way of specifying an architecture, some registers for threads, and some memory.  

I don't know if I want to structure it more generally yet, with subdirectories, or whatever, for the YAML to intermediate representation and for the Mach-O corefile writing.  If anyone does want to restructure it for a additional input/output methods, I think it will be easy to restructure it at that point.  It may end up never growing beyond this simple set of features (I know, unlikely)

jasonmolenda committed:
Integrate Jonas' suggestions.

jasonmolenda committed:
ws fix

jasonmolenda committed:
Add separate methods for writing 32-bit and 64-bit registers,
I think it's a little easier to follow the register context
writing methods when it's more explicit what is being written.

jasonmolenda committed:
Add the ability to specify a list of UUID & virtual addresses.
And the ability to specify the number of bits used in addressing
on this cpu.

jasonmolenda committed:
Merge branch 'main' into add-yaml2macho-core-test-utility

jasonmolenda committed:
ws fix

jasonmolenda committed:
Add YAML mappings so the number of addressable bits and binaries
can be specified directly in the .yaml file, while keeping the
command line options to augment/override those.

e.g.

```
cpu: arm64
addressable-bits:
    num-bits: 39

binaries:
  - name: binary-name.development
    uuid: 67942352-5857-3D3D-90CB-A3F80BA67B04
    virtual-address: 0xfffffff01840c000
```

jasonmolenda commented on lldb/test/API/macosx/riscv32-corefile/TestRV32MachOCorefile.py:
```diff
@@ -13,18 +13,17 @@
 
         process = target.GetProcess()
-        self.assertEqual(process.GetNumThreads(), 1)
+        self.assertEqual(process.GetNumThreads(), 2)
```
Yeah I was testing yaml2macho-core and needed more than one thread to make sure I was writing the load commands correctly (iirc I wasn't, when I had only one).

jasonmolenda commented:
> > You could imagine someone making an ELF corefile output capability, for instance.
> 
> Yes I'd love to be generating the core files eventually. Maybe we can with yaml2obj but I never seriously looked into it. As right now, patching in notes by hand makes for smaller files, but it's not reproducible or self documenting.
> 
> I assume the existing yaml2obj cannot make core files, or you wouldn't be writing this.

I don't think yaml2obj does make corefiles, but I believe this utility serves a different purpose.  yaml2obj could if you already have a small corefile (they're rarely small) that only contains the bytes needed to reproduce a bug that you want to test.  You may need to hand-modify an existing corefile to remove unnecessary memory segments, or to change values to elicit some behavior you want to test.

The idea with this tool is that you have the memory bytes that are critical to show an issue, or register values, and you want to construct a corefile based on those bytes. You may not be starting with a real corefile to begin with.  For that matter, you might construct stack memory values manually to elicit an unwinder bug, for instance.  (nb until I find a way for ObjectFileJSON to specify unwind rules for functions, this is unlikely to work, unless you copy the bytes of the relevant functions into the YAML so lldb can do its instruction emulation unwind plan).

Last week I had a bug report of someone seeing an issue with lldb in a firmware environment, where it would be difficult for me to set up the environment and corefiles were not an option.  I got the register values & 400 bytes of stack memory (lit. just "reg read" and "mem read"), and a copy of the relevant binary + dSYM, reformatted it all into YAML and created a Mach-O corefile that was able to reproduce the issue on my desktop.  Pretty handy!  Agree this is not common, but it was funny it came up only a couple weeks after I wrote the tool.

I think it fills a niche distinct from obj2yaml / yaml2obj, regardless of corefile support in that tool.

jasonmolenda committed:
ws

jasonmolenda committed:
ws fix, David's cleanup suggestion in the rv32 corefile api test.

jasonmolenda committed:
Need to include <optional> for linux.

jasonmolenda committed:
Make the second thread have unique register values
from the first thread.

jasonmolenda committed:
remove newlines between different include types.
Add brief descriptions of each header file's purpose as per
llvm/docs/CodingStandards.rst

jasonmolenda committed:
switch from the standard getopts to the less standard llvm

slydiman commented:
[lldb-remote-linux-win](https://lab.llvm.org/buildbot/#/builders/197/builds/8706) and [lldb-x86_64-win](https://lab.llvm.org/buildbot/#/builders/211/builds/1871) are still broken.

jasonmolenda commented:
Thanks.  I've asked someone with access to a Windows system to help me understand if this is an issue with my new yaml2macho-core utility (producing incorrect corefiles) or a problem with lldb on Windows reading (32-bit?) mach-o corefiles, I'll hold off on tweaking this any further for now.  If it's yaml2macho-core, the first thing I wonder is if I need add `b` to the fopen argument in `FILE *f = fopen(OutputFilename.c_str(), "w");` to indicate that binary is being written.  That's a no-op on unix systems but it might be relevant on Windows idk.
